### PR TITLE
Fix/json parsing groq

### PR DESCRIPTION
**Pull request by Varun2526**
## What does this PR do?

This PR improves how the project handles LLM responses, especially when using Groq models.

While running the fetch_mlas script, the LLM responses sometimes included extra text or markdown (like ```json blocks), which caused JSON parsing to fail and stopped the pipeline.

This PR fixes that by enforcing strict JSON output and improving the parsing logic so that such responses can still be handled correctly.

## Type of change

- [ ] Data contribution (added/updated MP or MLA records)
- [ ] New enrichment process (new agent process class)
- [x] Bug fix
- [x] Feature / enhancement
- [ ] Refactor / code cleanup
- [ ] Documentation update
- [ ] Other (describe below)

## Scope

- Updated JSON parsing logic in `app/agents/base_agent.py`
- Added handling for markdown code blocks and non-strict responses
- Improved prompts in `app/agents/state_mla_fetcher.py` to enforce JSON-only output
- Added fallback logic to safely extract JSON even if extra text is present

## How was this tested?

```bash
python3 scripts/fetch_mlas.py --state "Telangana" --log-level INFO

Result:
LLM responses were successfully parsed
No JSON parsing errors occurred
The full pipeline completed successfully
MLA data was fetched and stored correctly

## Version bump

<!-- Add ONE label to this PR to control the automatic version bump when merged to production. -->
<!-- If no label is added, defaults to `patch`. -->

- [✅ ] `patch` — Bug fixes, small tweaks (0.3.5 → 0.3.6)
- [ ] `minor` — New features, enhancements (0.3.5 → 0.4.0)
- [ ] `major` — Breaking changes (0.3.5 → 1.0.0)

## Checklist

- [ ✅] I have **not** committed `.env`, API keys, or any secrets
- [ ✅] I have **not** committed `app/database/cache.db`
- [ ✅] I have reviewed the diff and it only contains intended changes
- [ ✅] Data changes are limited to `app/data/mp.json` and/or `app/data/mla.json`
- [ ✅] Tests pass locally (`make test`)

## Screenshots / logs (optional)
![groq-success-run](https://github.com/user-attachments/assets/7ab33f5b-78b3-4980-ab9e-ab1553a9c6c9)


## Additional notes
Groq/LLaMA models often return responses with extra formatting or explanations. This PR ensures such outputs are handled correctly and improves overall reliability across different LLM providers

### CODE REVIEW

File: app/config/free_tier_llm.py [hunk 1]
```diff
@@ -102,8 +102,7 @@ def from_dict(cls, d: dict[str, Any]) -> ProviderConfig:
         "api_key_env": "GEMINI_API_KEY",
     },
     {"provider": "gemini", "model": "gemini-2.5-pro", "api_key_env": "GEMINI_API_KEY"},
-
-    {"provider": "gemini", "model": "gemini-3-flash", "api_key_env": "GEMINI_API_KEY"},
+    {"provider": "gemini", "model": "gemini-2.0-flash-lite", "api_key_env": "GEMINI_API_KEY"},
     {
```
[the remark this answers is not in the document]
Comment (Varun2526) [reply app/config/free_tier_llm.py:107]: Thanks for the suggestion!

The Gemini model change was part of an earlier fix and not directly related to this PR. I’ve now removed it to keep this PR focused only on JSON parsing improvements.

Let me know if anything else should be adjusted.

File: app/agents/base_agent.py [hunk 2]
```diff
@@ -224,15 +225,60 @@ def _run_llm_with_context(self, prompt: str, context: str) -> str:
...
+            except Exception:
+                pass
+
+        # 4. Use regex to find the outermost JSON object
+        object_match = re.search(r"(\{\s*\n?.*?\n?\s*\})", text, re.DOTALL)
+        if object_match:
+            try:
+                return json.loads(object_match.group(1))
+            except Exception:
+                pass
```
[the remark this answers is not in the document]
Comment (Varun2526) [reply app/agents/base_agent.py:280]: Thanks for the suggestion!

You're right that adding tests for these edge cases would improve reliability. 
For now, I’ve focused on fixing the parsing issue itself to keep the PR scope small.

I can follow up with a separate PR to add targeted tests for _parse_json_value.

File: app/agents/base_agent.py [hunk 3]
```diff
@@ -224,15 +225,60 @@ def _run_llm_with_context(self, prompt: str, context: str) -> str:
...
 
+        # 2. Strip markdown code fences and retry
+        stripped = self._strip_markdown_fences(text)
+        if stripped != text:
+            try:
+                return json.loads(stripped)
+            except Exception:
+                pass
+            # Continue with stripped text for bracket-matching below
+            text = stripped
```
[the remark this answers is not in the document]
Comment (Varun2526) [reply app/agents/base_agent.py]: Good catch — I’ve updated the logic to only use the stripped content when it successfully parses as JSON. 

This prevents losing valid data when non-JSON fenced blocks appear before the actual JSON. I also verified this with mixed fence scenarios and ensured the full pipeline still works correctly.

**Comment by PiyushLadukar** [comment]
@imsks Please review this PR

**Comment by imsks** [comment]
Hey @Varun2526 We shouldn't be using any Outdated models. We can't merge this PR as this doesn't server any specific purpose.

**Comment by Varun2526** [comment]
Hey @imsks, completely fair feedback — I should have made the purpose clearer upfront, that's on me.

Just to clarify though, this PR isn't about supporting outdated models. It's about making the pipeline bulletproof against a very common real-world issue — even state-of-the-art models like GPT-4o, Claude 3.5 Sonnet, or Groq's LLaMA-3 endpoints will occasionally wrap their output in markdown fences (```json ... ```) or add a sentence like "Here is the data you requested:" before the JSON.

When that happens even once, `fetch_mlas.py` crashes mid-run and the entire pipeline stops. Here's exactly what this PR adds as a safety net:

- `_strip_markdown_fences` in `base_agent.py` — strips ````json` blocks that LLMs add when mimicking ChatGPT-style formatting
- A regex extractor fallback — scans the full response and extracts only the `{...}` or `[...]` block, ignoring any conversational text
- Stricter prompts in `state_mla_fetcher.py` — explicitly tells the model to output ONLY raw JSON with no markdown or explanation
- Empty data fallback — if the model doesn't know, it returns `[]` instead of "I don't know", which would crash the parser

I tested this locally on Telangana and the full pipeline ran without a single JSON parsing error . This is  a reliability 
fix — model-agnostic and future-proof.

Either way, I'm a big fan of what you're building with Rajniti and genuinely want to contribute something useful. If this doesn't fit the current direction, I'll go through the open issues and find something that aligns better. Thanks for being straight with me! 🙌

**Comment by imsks** [comment]
Hey @Varun2526 Thanks. But this PR doesn't serve much help. Feel free to add a new change. We're bringing more changes where we need your help.
